File: app/src/main/java/org/stypox/dicio/skills/telephone/ContactChooserName.java
package org.stypox.dicio.skills.telephone;

import org.dicio.skill.Skill;
import org.dicio.skill.chain.InputRecognizer;
import org.stypox.dicio.util.StringUtils;

import java.util.Comparator;
import java.util.List;

public class ContactChooserName extends Skill {
    private final List<NameNumberPair> contacts;
    private String input;
    private NameNumberPair bestContact;

    ContactChooserName(final List<NameNumberPair> contacts) {
        this.contacts = contacts;
    }

    @Override
    public InputRecognizer.Specificity specificity() {
        // use a low specificity to prefer the index-based contact chooser
        return InputRecognizer.Specificity.low;
    }

    @Override
    public void setInput(final String theInput,
                         final List<String> inputWords,
                         final List<String> normalizedWordKeys) {
        this.input = theInput;
    }

    @Override
    public float score() {
        class Pair {
            final NameNumberPair nameNumberPair;
            final int distance;

            Pair(final NameNumberPair nameNumberPair, final int distance) {
                this.nameNumberPair = nameNumberPair;
                this.distance = distance;
            }
        }

        input = input.trim();
        bestContact = contacts.stream()
                .map(nameNumberPair -> new Pair(nameNumberPair,
                        StringUtils.customStringDistance(input, nameNumberPair.name)))
                .filter(pair -> pair.distance < 6)
                .min(Comparator.comparingInt(a -> a.distance))
                .map(pair -> pair.nameNumberPair)
                .orElse(null);

        return bestContact == null ? 0.0f : 1.0f;
    }

    @Override
    public void processInput() {
    }

    @Override
    public void generateOutput() {
        if (bestContact != null) {
            ConfirmCallOutput.callAfterConfirmation(this, bestContact.name, bestContact.number);
        }
    }

    @Override
    public void cleanup() {
        super.cleanup();
        bestContact = null;
    }
}

File: app/src/main/java/org/stypox/dicio/skills/telephone/ContactChooserName.kt
package org.stypox.dicio.skills.telephone

import org.dicio.skill.Skill
import org.dicio.skill.chain.InputRecognizer.Specificity
import org.stypox.dicio.util.StringUtils

class ContactChooserName internal constructor(private val contacts: List<NameNumberPair>) :
    Skill() {
    private var input: String? = null
    private var bestContact: NameNumberPair? = null
    override fun specificity(): Specificity {
        // use a low specificity to prefer the index-based contact chooser
        return Specificity.low
    }

    override fun setInput(
        theInput: String,
        inputWords: List<String>,
        normalizedWordKeys: List<String>
    ) {
        input = theInput
    }

    override fun score(): Float {
        class Pair(val nameNumberPair: NameNumberPair, val distance: Int)
        val input = input?.trim { it <= ' ' } ?: return 0.0f

        bestContact = contacts
            .map { nameNumberPair: NameNumberPair ->
                Pair(
                    nameNumberPair,
                    StringUtils.contactStringDistance(input, nameNumberPair.name)
                )
            }
            .filter { pair -> pair.distance < -7 }
            .minByOrNull { a -> a.distance }
            ?.nameNumberPair

        return if (bestContact == null) 0.0f else 1.0f
    }

    override fun processInput() {}
    override fun generateOutput() {
        bestContact?.also { ConfirmCallOutput.callAfterConfirmation(this, it.name, it.number) }
    }

    override fun cleanup() {
        super.cleanup()
        bestContact = null
    }
}